app/command.go

```diff
@@ -191,6 +191,100 @@ func (a *App) ExecuteCommand(args *model.CommandArgs) (*model.CommandResponse, *
 	return nil, model.NewAppError("command", "api.command.execute_command.not_found.app_error", map[string]interface{}{"Trigger": trigger}, "", http.StatusNotFound)
 }
 
+type mentionMapItem struct {
+	Name string
+	Id   string
+}
+
+// mentionsToTeamMembers returns all the @ mentions found in message that
+// belong to users in the specified team, linking them to their users
+func (a *App) mentionsToTeamMembers(message, teamId string) model.UserMentionMap {
+	possibleMentions := model.PossibleAtMentions(message)
+	mentionChan := make(chan *mentionMapItem, len(possibleMentions))
+
+	for _, mention := range possibleMentions {
+		go func(mention string) {
+			user, err := a.Srv.Store.User().GetByUsername(mention)
+			if err != nil {
+				// Consider trailing punctuation that may be hiding a valid username
+				trimmed, ok := model.TrimUsernameSpecialChar(mention)
+				for ; ok; trimmed, ok = model.TrimUsernameSpecialChar(trimmed) {
+					userFromTrimmed, userErr := a.Srv.Store.User().GetByUsername(trimmed)
+					if userErr != nil {
+						continue
+					}
+
+					_, err = a.GetTeamMember(teamId, userFromTrimmed.Id)
+					if err != nil {
+						// The user is not in the team, so we should ignore it
+						mentionChan <- nil
+						return
+					}
+
+					mentionChan <- &mentionMapItem{trimmed, userFromTrimmed.Id}
+					return
+				}
+
+				mentionChan <- nil
+				return
+			}
+
+			_, err = a.GetTeamMember(teamId, user.Id)
+			if err != nil {
+				// The user is not in the team, so we should ignore it
+				mentionChan <- nil
+				return
+			}
+
+			mentionChan <- &mentionMapItem{mention, user.Id}
+		}(mention)
+	}
+
+	atMentionMap := make(model.UserMentionMap)
+
+	for i := 0; i < len(possibleMentions); i++ {
```

Comment:
Synchronizing on `channels` and relying on right lengths between data structures always makes me a little nervous -- easy to forget a `mentionChan <- nil` in new code, for example.

Might we instead leverage `sync.WaitGroup` for each spawned goroutine, `wg.Wait()` for them all to finish, and then just consume whatever mentions are provided? This also eliminates the need to send a `nil` value into the channel at all.

Reply:
Ok, that seems much cleaner, thanks!

Reply:
Done in https://github.com/mattermost/mattermost-server/pull/13762/commits/552c67e6d459e98b646fd149c792a2bb0eb7486d. It looks much cleaner now, thank you for the suggestion :)